### ISSUE TITLE
NPE at io.intercom.android.sdk.Bridge.getContext(Bridge.java:104)

### ISSUE THREAD

**Issue by marcomorain** [intercom/intercom-android#215]
We've hit a crash in the Intercom SDK last night. It seems to the first time it's happened.

## Version info
  - Intercom for Android version: Not entirely sure. The app was built with `io.intercom.android:intercom-sdk:1.+@aar`
  - Android OS version: 6.0.1
  - Android Device: Samsung SM-G925F

## Steps to reproduce
This happened in the wild, so I'm not sure.

## Callstack

```
Fatal Exception: java.lang.RuntimeException: Unable to stop activity {ie.flipdish.the9thlough/ie.flipdish.flipdish_android.SplashActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.Context io.intercom.android.sdk.Bridge.context()' on a null object reference
       at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:5016)
       at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:5084)
       at android.app.ActivityThread.access$1700(ActivityThread.java:221)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1853)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:158)
       at android.app.ActivityThread.main(ActivityThread.java:7224)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1230)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1120)
Caused by java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.Context io.intercom.android.sdk.Bridge.context()' on a null object reference
       at io.intercom.android.sdk.Bridge.getContext(Bridge.java:104)
       at io.intercom.android.sdk.lifecycles.Lifecycles.isScreenLocked(Lifecycles.java:44)
       at io.intercom.android.sdk.lifecycles.Lifecycles.isApplicationInBackground(Lifecycles.java:57)
       at io.intercom.android.sdk.lifecycles.Lifecycles.onActivityStopped(Lifecycles.java:121)
       at android.app.Application.dispatchActivityStopped(Application.java:256)
       at android.app.Activity.onStop(Activity.java:1803)
       at android.support.v4.app.FragmentActivity.onStop(FragmentActivity.java:617)
       at android.support.v7.app.AppCompatActivity.onStop(AppCompatActivity.java:168)
       at android.app.Instrumentation.callActivityOnStop(Instrumentation.java:1306)
       at android.app.Activity.performStop(Activity.java:7070)
       at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:5011)
       at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:5084)
       at android.app.ActivityThread.access$1700(ActivityThread.java:221)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1853)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:158)
       at android.app.ActivityThread.main(ActivityThread.java:7224)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1230)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1120)
```

**Comment by ronocod** [comment]
That's very strange, the `Bridge` object is setup when you initialise the SDK and should never be null after that. Are you seeing a lot of instances of this crash?

**Comment by marcomorain** [comment]
Hi @ronocod, I've found 3 instances of this crash in Crashlytics, all with the same call-stack.

**Comment by marcomorain** [comment]
My knowledge of the Java memory model is rusty so take everything in this comment with a grain of salt.

Looking at the decompiled Bridge.java, I don't see any thread synchronization being performed on `Bridge.instance` - it's written to in a `synchronized` method but I don't believe this guarantees writes to the `instance` member to be visible to other threads immediately, unless the reading thread also calls a block synchronized on the same object (the reading thread would also have to call `Bridge.init()` to guarantee it has read the latest value of `instance`.

http://stackoverflow.com/a/21568663/249879

**Comment by marcomorain** [comment]
It looks like the application life-cycle callbacks require `Bridge.instance` to not be `null`, but the life-cycle callbacks are registered in the constructor of `Bridge`, at which point `Bridge.instance` is `null`, the Bridge singleton is only half-initialized. `Bridge.instance` is not set to a non-null value until after the Bridge constructor has returned. The application lifecycle callbacks can potentially fire at any point after they have been registered, which could be before `previewManager` has been assigned to.

I think registering for the callbacks should be deferred until after the Bridge object is fully initialised and `Bridge.instance` has been assigned to.

Current:

```Java
    Bridge(Application application, String apiKey, String appId) {
        this.appContext = application.getApplicationContext();
        this.poller = new PollingManager();
        this.bus = new MainThreadBus(ThreadEnforcer.ANY);
        this.superDeDuper = new DeDuper();
        this.nexusClient = new NexusWrapper();
        this.identityStore = new IdentityStore(this.appContext);
        this.metricsStore = new MetricsStore(this.appContext);
        this.api = new Api(this.appContext, appId, !this.identityStore.hasIntercomid());
        Lifecycles lifecycles = new Lifecycles(this.api, this.superDeDuper);
        application.registerActivityLifecycleCallbacks(lifecycles);
        this.previewManager = new PreviewManager(application, this.bus, this.poller, new PreviewDisplayManager());
        this.nexusClient.setLoggingEnabled(false);
        this.identityStore.setAppIdentity(apiKey, appId);
    }

    public static synchronized void init(Application application) {
        if(instance == null) {
            IntercomLogger.INTERNAL("bridge", "initialising bridge");
            IdentityStore idStore = new IdentityStore(application);
            initWithAppCredentials(application, idStore.getApiKey(), idStore.getAppId());
        }

    }

    public static synchronized void initWithAppCredentials(Application application, String apiKey, String appId) {
        if(instance == null) {
            IntercomLogger.INTERNAL("bridge", "initialising bridge");
            instance = new Bridge(application, apiKey, appId);
        }

    }
```

Proposed:

```Java
    Bridge(Application application, String apiKey, String appId) {
        this.appContext = application.getApplicationContext();
        this.poller = new PollingManager();
        this.bus = new MainThreadBus(ThreadEnforcer.ANY);
        this.superDeDuper = new DeDuper();
        this.nexusClient = new NexusWrapper();
        this.identityStore = new IdentityStore(this.appContext);
        this.metricsStore = new MetricsStore(this.appContext);
        this.api = new Api(this.appContext, appId, !this.identityStore.hasIntercomid());
        this.previewManager = new PreviewManager(application, this.bus, this.poller, new PreviewDisplayManager());
        this.nexusClient.setLoggingEnabled(false);
        this.identityStore.setAppIdentity(apiKey, appId);
    }

    public static synchronized void init(Application application) {
        if(instance == null) {
            IntercomLogger.INTERNAL("bridge", "initialising bridge");
            IdentityStore idStore = new IdentityStore(application);
            initWithAppCredentials(application, idStore.getApiKey(), idStore.getAppId());
        }

    }

    public static synchronized void initWithAppCredentials(Application application, String apiKey, String appId) {
        if(instance == null) {
            IntercomLogger.INTERNAL("bridge", "initialising bridge");
            instance = new Bridge(application, apiKey, appId);
            instance.registerLifecycleCallbacks();
        }

    }

  private void registerLifecycleCallbacks() {
    Lifecycles lifecycles = new Lifecycles(this.api, this.superDeDuper);
    application.registerActivityLifecycleCallbacks(lifecycles);
  }
```

Secondly, I think that any read of `private static Bridge instance` should be replaced with a `private static final AtomicReference<Bridge>` so that any read of the `instance` singleton in class `Bridge` can be guaranteed to read the latest value.

**Comment by ronocod** [comment]
This is great, we'll add in this change and I'll update this thread when it's released. Thanks!

**Comment by ronocod** [comment]
Thanks for the help with this, we've integrated your fixes into [3.0.0](https://github.com/intercom/intercom-android/releases/tag/3.0.0) which is out now. Hopefully this will fix your issue 😄

**Comment by marcomorain** [comment]
Thanks @ronocod
